02_helper_functions/AutomatedHallCalibration.m

```diff
@@ -44,76 +44,149 @@
-        
-        % reset recording state
-        recording_step = false;
-        
-    else
-        % we are within the deviation limits for successive measurements
-        
-        if (~recording_step)
-            % check if current data is sufficiently far from previously
-            % logged step mean (this check prevents repeat measurements
-            % from same angle)
-            if (i_cal == 0)
-                % protect agains 0 index on first step
-                last_mean = 0;
-            else
-                last_mean = cal_data(i_cal, 2);
+        if recording_step
+            % check if step is detected or end of data
+            if std_dev(k) > cal_opt.std_dev_threshold  | (k == idx_st+len_t-1)
+                if k > offset
+                    if sensor_hall_mag_T.Time(k - offset) - sensor_hall_mag_T.Time(k_start) > cal_opt.t_step_min
+                        % set current calibration step
+                        i_cal = i_cal + 1;
+
+                        % fill in data
+                        cal_data(i_cal, 2) = ...
+                            mean(sensor_hall_mag_T.Data(k_start:(k - offset)));
+
+                        cal_data(i_cal, 3) = ...
+                            std(sensor_hall_mag_T.Data(k_start:(k - offset)));
+
+                        cal_data(i_cal, 4) = k_start;
+                        cal_data(i_cal, 5) = k - offset;
+
+                        last_mean = cal_data(i_cal, 2);
+                    else
+                        disp('Measurement step rejected because it was too short')
+                    end
+                else
+                    error('The first calibration step was not long enough')
+                end
+
+                recording_step = false;
             end
-            if (abs(sensor_hall_mag_T.Data(k) - last_mean) > cal_opt.step_thres || i_cal == 0)
-                % start recording step
-                recording_step = true;
-                cal_data(i_cal+1, 4) = k;
+        else
+            % check if measurements stabilise
+            if std_dev(k) < cal_opt.std_dev_threshold
+                if abs(averaged(k) - last_mean) > cal_opt.step_thres
+                    recording_step = true;
+                    k_start = k + offset;
+                end
             end
         end
-        
-        if (recording_step)
-            % time since started recording
-            t_step = sensor_hall_mag_T.Time(k) - sensor_hall_mag_T.Time(cal_data(i_cal+1, 4));
-        end
     end
+    figure()
+    plot(sensor_hall_mag_T.Time, averaged)
+    ylabel('Filtered Measurements [mT]')
+    xlabel('Time [s]')
+    figure()
+    plot(sensor_hall_mag_T.Time, std_dev)
+    ylabel('Standard Deviation of Filtered Measurements [mT]')
+    xlabel('Time [s]')
+else
+    error('Unkown step detection mode: ' + string(cal_opt.step_detection_mode))
 end
 
+if i_cal ~= len_cal_data
+    error('Only ' + string(i_cal) + '/' + string(len_cal_data) + ' steps were detected')
```

Comment:
this message didnt work for me unless i type converted the content in the parentheses to char(...). but i have a very old version of matlab... 2016b... so also reasonable if you think I should just update and we leave it ;)

Reply:
Well, I think some backward compatibility is nice but I think with Matlab updating usually works quite well. I am currently running 2018 but probably will update it soon.
Anyway this is also only an error message so if it fails it is not the biggest harm.

Reply:
ok for me